src/clj/puppetlabs/services/master/file_serving.clj

```diff
@@ -166,3 +303,62 @@
 
       :else
       (ruby-request-handler request))))
+
+(defn metadatas-params-errors
+  "Reject requests with parameter values that we haven't implemented yet.
+  Returns a list of errors."
+  [params]
+  (->>
+   (for [[value default] {"recurse" "true"
+                          "links" "follow"
+                          "source_permissions" "ignore"
+                          "recurselimit" "infinite"}]
+     (when (and (get params value)
+                (not (= value default)))
+       (str "The only supported value of `" value "` at this time is `" default "`")))
+   (filter (comp not nil?))))
+
+(defn handle-project-file-metadatas
+  "Handle a file_metadatas request for a bolt project."
+  [bolt-projects-dir request]
+  (let [project (get-in request [:params "project"])
+        path (get-in request [:params :rest])
+        match (bidi/match-route project-routes path)]
+    (when match
+      (let [mount-point (get-in match [:route-params :mount-point])
+            mount-type (get-in match [:handler])
+            ignore (get-in request [:params "ignore"] [])
+            checksum-type (get-in request [:params "checksum_type"] "md5")]
```

Comment:
I was wondering about this...

I think in practice we always get the checksum type specified by the client, but also that puppet is still using md5 here.

Reply:
Okay. I think that changed in Puppet 7. @joshcooper can you confirm?

Reply:
Yeah, when the [server creates the metadata object](https://github.com/puppetlabs/puppet/blob/f664d6a21656f80b54b396529c84ee7e24108921/lib/puppet/file_serving/terminus_helper.rb#L9) it defaults to the [server's `Puppet[:digest_algorithm]` value](https://github.com/puppetlabs/puppet/blob/f664d6a21656f80b54b396529c84ee7e24108921/lib/puppet/file_serving/metadata.rb#L135). But if the client requests a specific checksum, then [it takes precedence](https://github.com/puppetlabs/puppet/blob/f664d6a21656f80b54b396529c84ee7e24108921/lib/puppet/file_serving/terminus_helper.rb#L12), so that the call to `FileMetadata#collect` uses that.

Puppet 7 does default to `sha256`, but it could be switched to `md5` I think the only reason to do that would be if you're using remote file buckets and you have pre-7 agents.

Reply:
This is all new stuff, I'll make it default to sha256

Reply:
However, I do think we want this targeted at 6.x, right? Everything else has landed in that stream so far.